dig.go

```diff
@@ -352,13 +352,24 @@ func (c *Container) remove(nodes []*node) {
 }
 
 func (c *Container) constructorArgs(ctype reflect.Type) ([]reflect.Value, error) {
+	numArgs := ctype.NumIn()
+
 	args := make([]reflect.Value, 0, ctype.NumIn())
-	for i := 0; i < ctype.NumIn(); i++ {
+	for i := 0; i < numArgs; i++ {
 		arg, err := c.get(edge{key: key{t: ctype.In(i)}})
 		if err != nil {
 			return nil, fmt.Errorf("couldn't get arguments for constructor %v: %v", ctype, err)
 		}
-		args = append(args, arg)
+
+		// If the function is variadic, the last argument is a slice that must
+		// be expanded into the argument list.
+		if i == numArgs-1 && ctype.IsVariadic() {
+			for j := 0; j < arg.Len(); j++ {
```

Comment:
I think there is a distinction between them: [varargs are passed as slices](https://golang.org/ref/spec#Passing_arguments_to_..._parameters), but they are not slices. If someone wants to pass slice elements as args, they need to expand it by adding `...` after slice.
```golang
package main

func print(nums ...int) {
	for _, n := range nums {
		println(n)
	}
}

func main() {
	print([]int{3,2,1}...)
	println("now print argument without slices:")
	print(7)
	print(6,5,4)
}
```

Which leads to a following case: if we have variable x and y (`var T x` and `var []T y`) in the graph, how should we handle invoke of of some `f(v ...T)`? I argue that we should use `x`, not `y`, because in your program you can write `f(x)`. `f(y)` will lead to compilation error, because you need **explicitly** expand y: `f(y...)`.

My suggestion is based on the above argument and I see varargs as optionals in the end of argument list for a function, not as slice from the container. Dig allows us to store only one instance per type, that's why I think the length could be only 0 or 1. I don't think that we can pass named instances as slices, because names carry important information about instances and can't be blindly applied.